core/amber/src/main/scala/edu/uci/ics/texera/web/resource/GmailResource.scala

```diff
@@ -0,0 +1,122 @@
+  final private lazy val clientSecret =
+    AmberUtils.amberConfig.getString("user-sys.google.clientSecret")
+
+  @GET
+  @RolesAllowed(Array("ADMIN"))
+  @Path("/email")
+  def getEmail: String = {
+    new String(Files.readAllBytes(path.resolve("userEmail")))
+  }
+
+  @POST
+  @RolesAllowed(Array("ADMIN"))
+  @Path("/auth")
+  def saveRefreshToken(code: String): Unit = {
+    this.deleteEmail()
+    val token = new GoogleAuthorizationCodeTokenRequest(
+      new NetHttpTransport(),
+      new GsonFactory(),
+      clientId,
+      clientSecret,
+      code,
+      "postmessage"
+    ).execute()
+
+    Files.write(
+      path.resolve("refreshToken"),
+      token.getRefreshToken.getBytes
+    )
+
+    Files.write(
+      path.resolve("userEmail"),
+      new GoogleIdTokenVerifier.Builder(new NetHttpTransport, GsonFactory.getDefaultInstance)
+        .setAudience(Collections.singletonList(clientId))
+        .build()
+        .verify(token.getIdToken)
+        .getPayload
+        .getEmail
+        .getBytes
+    )
+  }
+
+  @DELETE
+  @RolesAllowed(Array("ADMIN"))
+  @Path("/revoke")
+  def deleteEmail(): Unit = {
+    if (Files.exists(path.resolve("userEmail"))) Files.delete(path.resolve("userEmail"))
+    if (Files.exists(path.resolve("refreshToken"))) Files.delete(path.resolve("refreshToken"))
+    if (Files.exists(path.resolve("accessToken"))) Files.delete(path.resolve("accessToken"))
+  }
+
+  @PUT
+  @RolesAllowed(Array("REGULAR", "ADMIN"))
+  @Path("/send")
+  def sendEmail(emailMessage: EmailMessage, @Auth user: SessionUser): Unit = {
+    val accessTokenPath = path.resolve("accessToken")
+    if (!Files.exists(accessTokenPath)) {
+      Files.write(
+        accessTokenPath,
+        new GoogleRefreshTokenRequest(
+          new NetHttpTransport(),
+          new GsonFactory(),
+          new String(Files.readAllBytes(path.resolve("refreshToken"))),
+          clientId,
+          clientSecret
+        ).execute().getAccessToken.getBytes
+      )
+    }
+    val accessToken = new String(Files.readAllBytes(accessTokenPath))
+    val credentials = GoogleCredentials.create(new AccessToken(accessToken, null))
+    val service = new Gmail.Builder(
+      new NetHttpTransport(),
+      GsonFactory.getDefaultInstance,
+      new HttpCredentialsAdapter(credentials)
+    ).setApplicationName("Gmail samples").build()
+    val email = new MimeMessage(Session.getDefaultInstance(new Properties(), null))
+    email.setFrom(new InternetAddress("me"))
+    email.addRecipient(
+      javax.mail.Message.RecipientType.TO,
+      new InternetAddress(if (emailMessage.email == "") user.getEmail else emailMessage.email)
+    )
+    email.setSubject(emailMessage.subject)
+    email.setText(emailMessage.content)
+    val buffer = new ByteArrayOutputStream()
+    email.writeTo(buffer)
+    val message = new Message()
+    message.setRaw(Base64.encodeBase64URLSafeString(buffer.toByteArray))
+    service.users().messages().send("me", message).execute()
```

Comment:
It looks like the `send` api only sends the email and does not have any response to the frontend? I'm not sure if this is a good idea. Is there a way to verify that the email is successfully sent?

Reply:
Will show notification on frontend if there is a error.